### DISCUSSION TITLE
tutorials/voyage-part2/09-using-override

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#172] [course feedback]
# tutorials/voyage-part2/09-using-override

https://www.apollographql.com/tutorials/voyage-part2/09-using-override

**Comment by sheilaklotz** [comment]
My rover dev is watching both subgraph-accounts\schema.graphql and monolith\schema.graphql. Should it show the errors automaticaly? My rover dev does nothing when I save the schema files. It just shows errors if I stop it and start it again. Is this the expected behavior?

**Reply by mabuyo** [reply]
Hi @sheilaklotz ! Whenever you modify one of those subgraph schemas and save your changes, Rover should automatically re-compose and surface any errors. Could you make sure those files were saved? What messages is Rover showing?

**Reply by sheilaklotz** [reply]
Yeah, the files were saved multiple times. No message, even if should show an error.

And when I test the router, it can bring data from monolith schema, but not from accounts one.

When I try this query
uery User($userId: ID!) {
  user(id: $userId) {
    ... on Host {
      overallRating
    }
  }
}

the response is: 
{
  "data": {
    "user": {
      "overallRating": 4.166666666666667
    }
  }
}

So, it can resolve the overallRating field.

When I try this query:
query User($userId: ID!) {
  user(id: $userId) {
    ... on Host {
      overallRating
      name
    }
  }
}

I get an error, because it cannot resolve it
{
  "data": {
    "user": null
  },
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Host.name.",
      "path": [
        "user",
        "name"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "Error: Cannot return null for non-nullable field Host.name.",
          "    at completeValue (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:605:13)",
          "    at executeField (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:500:19)",
          "    at executeFields (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:414:22)",
          "    at completeObjectValue (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:925:10)",
          "    at completeAbstractValue (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:806:10)",
          "    at completeValue (C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:635:12)",
          "    at C:\\Users\\SheilaKlotz\\Documents\\Apollo GraphQL\\Voyage II - Federating the Monolith\\odyssey-voyage-II-server\\subgraph-accounts\\node_modules\\graphql\\execution\\execute.js:697:11",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async Promise.all (index 0)",
          "    at async Promise.all (index 0)"
        ]
      }
    }
  ],
  "extensions": {
    "valueCompletion": [
      {
        "message": "Cannot return null for non-nullable field Host.name",
        "path": [
          "user"
        ]
      }
    ]
  }
}

**Reply by mabuyo** [reply]
Hmm, it looks like the router might not be picking up the accounts subgraph. A couple things to check: 
- What's in your `supergraph-config.yaml` file?
- Is the `accounts` subgraph running, on `localhost:4002`?
I'm also happy to try and troubleshoot if you have your code pushed up so I can try to run and reproduce it!

**Reply by sheilaklotz** [reply]
subgraph accounts running

[nodemon] restarting due to changes...
[nodemon] starting `node index.js`
Enabling inline tracing for this subgraph. To disable, use ApolloServerPluginInlineTraceDisabled.
🚀 Subgraph accounts running at http://localhost:4002/


supergraph-config.yaml
federation_version: =2.7.0
subgraphs:
  monolith:
    routing_url: http://localhost:4001
    schema:
      file: ./monolith/schema.graphql
  accounts:
    routing_url: http://localhost:4002
    schema:
      file: ./subgraph-accounts/schema.graphql

Sorry, how can I have my code pushed up?

**Reply by mabuyo** [reply]
That code looks right! By pushed up I mean committing your changes and pushing them up to GitHub. I see that you have the repository forked here https://github.com/sheilaklotz/odyssey-voyage-II-server -- would you be able to update the repo with your changes?

**Reply by sheilaklotz** [reply]
Sorry for the delay. I pushed my changes to my fork.

**Reply by at1217** [reply]
The issue arises because we have defined the `name` field as non-null by specifying it as a `String `datatype followed by the `!` operator. However, we have not yet implemented a resolver to handle this field, which causes the accounts subgraph to return `null` for it. By removing the non-null constraint from the `name` field in both the `User` interface and the `Host` type, the error will be resolved. Once the resolver is implemented in the following section, the subgraph will consistently return a value for the `name` field, allowing us to eventually redefine it as non-null.

**Reply by mabuyo** [reply]
Thank you for the clarification @at1217! 

@sheilaklotz - apologies for following up on this so late. We do expect the response _data_ to have errors at this stage. However in this lesson, we aren't looking at the response data yet, we are just looking at the query plan to demonstrate what happens when we we use `@override`. The overall schema _should_ still compose with no errors at this point.

The next lesson will address the _data_ errors by implementing resolvers.

**Comment by stephan-buckmaster** [comment]
I seem to have the same problem as @sheilaklotz oh well.

**Comment by giovai** [comment]
Hello, when I try to test the progressive override on `updateProfile` Mutation, I get errors:
```
type Mutation {
   "Updates the logged-in user's profile information"
  updateProfile(updateProfileInput: UpdateProfileInput): UpdateProfileResponse! @override(from: "monolith", label: "percent(25)")
}
```

ERRORS:
```
ATISFIABILITY_ERROR: The following supergraph API query:
    mutation {
      updateProfile(updateProfileInput: {}) {
        code
      }
    }
    cannot be satisfied by the subgraphs because:
    - from subgraph "monolith":
      - field "UpdateProfileResponse.code" is not resolvable because it is overridden by subgraph "accounts".
      - cannot move to subgraph "accounts", which has field "UpdateProfileResponse.code", because type "UpdateProfileResponse" has no @key defined in subgraph "accounts”.
```

Same error on all `UpdateProfileResponse` fields.
```
"The response after updating a profile"
type UpdateProfileResponse implements MutationResponse {
  "Similar to HTTP status code, represents the status of the mutation"
  code: Int! @override(from: "monolith")
  "Indicates whether the mutation was successful"
  success: Boolean! @override(from: "monolith")
  "Human-readable message for the UI"
  message: String! @override(from: "monolith")
  "Updated user"
  user: User @override(from: "monolith")
}
```
